elijah/deque-template.js

```diff
@@ -0,0 +1,171 @@
+// Problem 2a): build a deque factory
+//-------
+
+// The factory itself:
+function makeDeque(stuff) {
+    // return deque object
+    return {
+        values: stuff,
+        // slice to make sure not linked
+        deque: stuff.slice(),
+        dropped: [],
+        arrLength: makeDeque.arrLength,
+        top: makeDeque.top,
+        bottom: makeDeque.bottom,
+        pop: makeDeque.pop,
+        push: makeDeque.push,
+        shift: makeDeque.shift,
+        unshift: makeDeque.unshift,
+        cut: makeDeque.cut,
+        map: makeDeque.map,
+        sort: makeDeque.sort,
+        shuffle: makeDeque.shuffle,
+        shuffleProperly: makeDeque.shuffleProperly,
+        // checks for valid values
+        dequeCheck: makeDeque.dequeCheck
+    }
+
+}
+
+// The factory's instance methods:
+makeDeque.arrLength = function() {
+    // return lenght of "deque"
+	return this.values.length;
+}
+
+makeDeque.top = function() {
+    // return top of deque
+	return this.values[0];
+}
+
+makeDeque.bottom = function() {
+    // return "bottom" of deque
+	return this.values[this.values.length -1];
+}
+
+makeDeque.pop = function() {
+	// pop the top value off and return
+    var p = this.values.splice(0,1)[0];
+    this.dropped.push(p);
+    return p;
+}
+
+makeDeque.push = function(val) {
+	// add val to top of values and return new length
+    if (this.dequeCheck(val)) {
+        this.values.splice(0,0,val);
+        return this.arrLength();
+    } else {
+        return null;
+    }
+}
+
+makeDeque.shift = function() {
+	// remove and return bottom
+    var s = this.values.splice((this.values.length -1),1)[0];
+    this.dropped.push(s);
+    return s;
+}
+
+makeDeque.unshift = function(val) {
+	// add val to bottom and return new length
+    if (this.dequeCheck(val)) {
+        this.values.splice((this.values.length),0,val);
+        return this.arrLength();
+    } else {
+        return null;
+    }
+}
+
+makeDeque.cut = function() {
+	// split deque in half and swap them
+    var mid = Math.round(this.arrLength()/2);
+    var a = this.values.slice(mid);
+    var b = this.values.slice(0,mid);
+    this.values = a.concat(b);
+}
+
+makeDeque.map = function(convert) {
+	// return array of elements from callback
+    return this.values.map(convert);
+}
+
+makeDeque.sort = function(compare) {
+	// reorder deque values with callback
+    this.values.sort(compare);
+}
+
+makeDeque.shuffle = function() {
+    // shuffle cards in values using ineffective random way
+    // use our sort method
+    this.sort(badShuffle);
+}
+
+makeDeque.shuffleProperly = function() {
+    // shuffle cards properly using in-place Knuth-Fisher-Yates
+    // http://bost.ocks.org/mike/shuffle/
+    var len = this.values.length, t, i;
+    while (len) {
+        i = Math.floor(Math.random() * len--);
+        t = this.values[len];
+        this.values[len] = this.values[i];
+        this.values[i] = t;
+    }
+}
+
+// 2e checker function
+makeDeque.dequeCheck = function(val) {
```

Comment:
2e looks great! Returns null when I try to add items that are not in the dropped array of deckofcards, and is giving me a new length when adding back in. Good work.